overlord/snapshotstate/snapshotstate_test.go

```diff
@@ -744,6 +746,122 @@ exec /bin/tar "$@"
 	c.Check(string(out), check.Equals, "")
 }
 
+func (snapshotSuite) testSaveIntegrationTarFails(c *check.C, tarLogLines int, expectedErr string) {
+	if os.Geteuid() == 0 {
+		c.Skip("this test cannot run as root (runuser will fail)")
+	}
+	c.Assert(os.MkdirAll(dirs.SnapshotsDir, 0755), check.IsNil)
+	// sanity check: no files in snapshot dir
+	out, err := exec.Command("find", dirs.SnapshotsDir, "-type", "f").CombinedOutput()
+	c.Assert(err, check.IsNil)
+	c.Check(string(out), check.Equals, "")
+
+	homedir := filepath.Join(dirs.GlobalRootDir, "home", "a-user")
+	// mock tar so that it outputs a desired number of lines
+	tarFailScript := fmt.Sprintf(`
+export LANG=C
+for c in $(seq %d); do echo "log line $c" >&2 ; done
+exec /bin/tar "$@"
+`, tarLogLines)
+	if tarLogLines == 1 {
+		tarFailScript = "echo nope >&2 ; exit 1"
+	} else if tarLogLines == 0 {
+		tarFailScript = "exit 1"
+	}
+	mocktar := testutil.MockCommand(c, "tar", tarFailScript)
+	defer mocktar.Restore()
+
+	defer backend.MockUserLookup(func(username string) (*user.User, error) {
+		if username != "a-user" {
+			c.Fatalf("unexpected user %q", username)
+		}
+		return &user.User{
+			Uid:      fmt.Sprint(sys.Geteuid()),
+			Username: username,
+			HomeDir:  homedir,
+		}, nil
+	})()
+
+	o := overlord.Mock()
+	st := o.State()
+
+	stmgr, err := snapstate.Manager(st, o.TaskRunner())
+	c.Assert(err, check.IsNil)
+	o.AddManager(stmgr)
+	shmgr := snapshotstate.Manager(st, o.TaskRunner())
+	o.AddManager(shmgr)
+	o.AddManager(o.TaskRunner())
+
+	st.Lock()
+	defer st.Unlock()
+
+	sideInfo := &snap.SideInfo{RealName: "tar-fail-snap", Revision: snap.R(1)}
+	snapstate.Set(st, "tar-fail-snap", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{sideInfo},
+		Current:  sideInfo.Revision,
+		SnapType: "app",
+	})
+	snaptest.MockSnap(c, "name: tar-fail-snap\nversion: v1", sideInfo)
+	c.Assert(os.MkdirAll(filepath.Join(homedir, "snap/tar-fail-snap/1/canary-tar-fail-snap"), 0755), check.IsNil)
+	c.Assert(os.MkdirAll(filepath.Join(homedir, "snap/tar-fail-snap/common"), 0755), check.IsNil)
+	// this makes tar unhappy
+	c.Assert(os.Mkdir(filepath.Join(homedir, "snap/tar-fail-snap/common/common-tar-fail-snap"), 00), check.IsNil)
+
+	setID, saved, taskset, err := snapshotstate.Save(st, nil, []string{"a-user"})
+	c.Assert(err, check.IsNil)
+	c.Check(setID, check.Equals, uint64(1))
+	c.Check(saved, check.DeepEquals, []string{"tar-fail-snap"})
+
+	change := st.NewChange("save-snapshot", "...")
+	change.AddAll(taskset)
+
+	st.Unlock()
+	c.Assert(o.Settle(5*time.Second), check.IsNil)
```

Comment:
(nitpick^2) Maybe use `var settleTimeout = testutil.HostScaledTimeout(5 * time.Second)` on the top and use settleTimeout here? Not sure if we really need the host scaled timeout, 5s seems eternity for creating a tar - but then we also have risc-v :)